src/storage/mod.rs

```diff
@@ -129,6 +140,145 @@ impl Storage {
+            date_updated: blob.date_updated,
+            content: blob.content,
+            compression: None,
+        })
+    }
+
+    pub(crate) fn store_all_in_archive(
+        &self,
+        archive_path: &str,
+        root_dir: &Path,
+    ) -> Result<(HashMap<PathBuf, String>, CompressionAlgorithm), Error> {
+        let mut file_paths = HashMap::new();
+
+        // due to how we retrieve the archived files, the compression method has to be supported
+        // by our normal content-encoding / compression logic.
+        // Deflate is not supported right now, and is compressing worse anyways.
+        let options =
+            zip::write::FileOptions::default().compression_method(zip::CompressionMethod::Bzip2);
+
+        let mut zip = zip::ZipWriter::new(io::Cursor::new(Vec::new()));
+        for file_path in get_file_list(root_dir)? {
+            let mut file = fs::File::open(root_dir.join(&file_path))?;
+
+            // TODO: should we create directories with `start_directory`?
+            // it's not standard, but convention for zip files
+
+            zip.start_file(file_path.to_str().unwrap(), options)?;
+            io::copy(&mut file, &mut zip)?;
+
+            let mime = detect_mime(&file_path);
+            file_paths.insert(file_path, mime.to_string());
+        }
+
+        let mut zip_content = zip.finish()?.into_inner();
+        let index = archive_index::Index::new_from_zip(&mut io::Cursor::new(&mut zip_content))?;
+        let mut index_content = vec![];
+        index.save(&mut index_content)?;
+        let alg = CompressionAlgorithm::default();
+        let compressed_index_content = compress(&index_content[..], alg)?;
+
+        let remote_index_path = format!("{}.index", &archive_path);
+
+        // additionally store the index in the local cache, so it's directly available
+        let local_index_path = self.local_archive_cache_path.join(&remote_index_path);
+        if local_index_path.exists() {
+            fs::remove_file(&local_index_path)?;
+        }
+        fs::create_dir_all(local_index_path.parent().unwrap())?;
+        let mut local_index_file = fs::File::create(&local_index_path)?;
+        local_index_file.write_all(&index_content)?;
+
+        self.store_inner(
+            vec![
+                Blob {
+                    path: archive_path.to_string(),
+                    mime: "application/zip".to_owned(),
+                    content: zip_content,
+                    compression: None,
+                    date_updated: Utc::now(),
+                },
+                Blob {
+                    path: remote_index_path,
+                    mime: "application/octet-stream".to_owned(),
+                    content: compressed_index_content,
+                    compression: Some(alg),
+                    date_updated: Utc::now(),
+                },
+            ]
+            .into_iter()
+            .map(Ok),
+        )?;
+
+        Ok((file_paths, CompressionAlgorithm::Bzip2))
```

Comment:
nit: I would add a `file_alg` local variable near the top so it's clear this may be different from the compression of the index.

Why *does* the index use a different compression, by the way?

Reply:
> Why _does_ the index use a different compression, by the way?

The index-file is just another file. So I chose the better algorithm for compression.

Reply:
I guess my question is when you'd expect the default algorithm to not have the best compression. If they're different maybe we should specify the algorithm explicitly instead of just saying ::default()?